File: python/src/jevlint/self_test_formatter.py
"""What each check scored against its own clean and broken examples."""
from __future__ import annotations

from .formatting import dim, fixed, pad, pad_left, paint
from .selftest import CheckScore
from .text import Text


class SelfTestFormatter:
    def __init__(self, colour: bool = True) -> None:
        self._colour = colour

    def format(self, scores: list[CheckScore]) -> str:
        def name(score: CheckScore) -> str:
            return score.check.id if score.domain == '' else score.check.id + ' (' + score.domain + ')'

        # Wide enough for the longest id with its domain, so a long one does not push its row out of line.
        width = max([44, *(len(name(score)) for score in scores)])
        lines = [
            dim(self._colour, Text.of('self_test.heading')),
            '',
            paint(self._colour, _row(
                width,
                Text.of('self_test.column_check'),
                Text.of('self_test.column_clean'),
                Text.of('self_test.column_broken'),
                Text.of('self_test.column_fixed'),
                Text.of('self_test.column_span'),
                Text.of('self_test.column_verdict'),
            ), '1'),
        ]

        for score in scores:
            lines.append(_row(
                width,
                name(score),
                _number(score.clean),
                _number(score.broken),
                _number(score.fixed),
                _number(score.span()),
                self._verdict(score),
            ))

        failed = [score for score in scores if not score.passed()]
        lines.append('')
        # One row per example set, and every check ships two. Counting rows as
        # checks read as though `--check=one-id` had matched two ids.
        checks = {score.check.id for score in scores}
        lines.append(Text.of('self_test.summary', {
            'separated': len(scores) - len(failed),
            'sets': len(scores),
            'checks': len(checks),
        }))

        if len(failed) > 0:
            lines.append(dim(self._colour, Text.of('self_test.footer')))

        return '\n'.join(lines) + '\n'

    def _verdict(self, score: CheckScore) -> str:
        verdict = score.verdict()

        if verdict == 'ok':
            return paint(self._colour, verdict, '32')

        return paint(self._colour, verdict, '33' if verdict == 'weak' else '31')


def _row(width: int, check: str, clean: str, broken: str, fix: str, span: str, verdict: str) -> str:
    return (
        pad(check, width) + ' ' + pad_left(clean, 7) + ' ' + pad_left(broken, 7) + ' '
        + pad_left(fix, 7) + ' ' + pad_left(span, 7) + '  ' + verdict
    )


def _number(value: float | None) -> str:
    return 'n/a' if value is None else fixed(value, 2)

File: js/src/format/selfTestFormatter.ts
import { Text } from '../i18n/text.js';
import type { CheckScore } from '../selftest/checkScore.js';
import { dim, pad, padLeft, paint } from './colour.js';
import { fixed } from './numbers.js';

/** What each check scored against its own clean and broken examples */
export class SelfTestFormatter {
    constructor(private readonly colour = true) {}

    format(scores: CheckScore[]): string {
        const name = (score: CheckScore): string => (score.domain === '' ? score.check.id : `${score.check.id} (${score.domain})`);
        // Wide enough for the longest id with its domain, so a long one does not push its row out of line.
        const width = Math.max(44, ...scores.map((score) => name(score).length));
        const lines = [
            dim(this.colour, Text.of('self_test.heading')),
            '',
            paint(this.colour, row(
                width,
                Text.of('self_test.column_check'),
                Text.of('self_test.column_clean'),
                Text.of('self_test.column_broken'),
                Text.of('self_test.column_fixed'),
                Text.of('self_test.column_span'),
                Text.of('self_test.column_verdict'),
            ), '1'),
        ];

        for (const score of scores) {
            lines.push(row(
                width,
                name(score),
                number(score.clean),
                number(score.broken),
                number(score.fixed),
                number(score.span()),
                this.verdict(score),
            ));
        }

        const failed = scores.filter((score) => !score.passed());
        lines.push('');
        // One row per example set, and every check ships two. Counting rows as
        // checks read as though `--check=one-id` had matched two ids.
        const checks = new Set(scores.map((score) => score.check.id));
        lines.push(Text.of('self_test.summary', {
            separated: scores.length - failed.length,
            sets: scores.length,
            checks: checks.size,
        }));

        if (failed.length > 0) {
            lines.push(dim(this.colour, Text.of('self_test.footer')));
        }

        return `${lines.join('\n')}\n`;
    }

    private verdict(score: CheckScore): string {
        const verdict = score.verdict();

        if (verdict === 'ok') {
            return paint(this.colour, verdict, '32');
        }

        return paint(this.colour, verdict, verdict === 'weak' ? '33' : '31');
    }
}

function row(width: number, check: string, clean: string, broken: string, fix: string, span: string, verdict: string): string {
    return `${pad(check, width)} ${padLeft(clean, 7)} ${padLeft(broken, 7)} ${padLeft(fix, 7)} ${padLeft(span, 7)}  ${verdict}`;
}

function number(value: number | null): string {
    return value === null ? 'n/a' : fixed(value, 2);
}
